cpp/src/strings/char_types/char_cases.h

```diff
@@ -678,3 +678,204 @@ const uint16_t g_character_cases_table[] = {
+   { 2, { 7980, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 7981, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 7982, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 7983, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 7976, 921, 0 }, 1, {8080, 0, 0} },
+   { 2, { 7977, 921, 0 }, 1, {8081, 0, 0} },
+   { 2, { 7978, 921, 0 }, 1, {8082, 0, 0} },
+   { 2, { 7979, 921, 0 }, 1, {8083, 0, 0} },
+   { 2, { 7980, 921, 0 }, 1, {8084, 0, 0} },
+   { 2, { 7981, 921, 0 }, 1, {8085, 0, 0} },
+   { 2, { 7982, 921, 0 }, 1, {8086, 0, 0} },
+   { 2, { 7983, 921, 0 }, 1, {8087, 0, 0} },
+   { 2, { 8040, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 8041, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 8042, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 8043, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 8044, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 8045, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 8046, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 8047, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 8040, 921, 0 }, 1, {8096, 0, 0} },
+   { 2, { 8041, 921, 0 }, 1, {8097, 0, 0} },
+   { 2, { 8042, 921, 0 }, 1, {8098, 0, 0} },
+   { 2, { 8043, 921, 0 }, 1, {8099, 0, 0} },
+   { 2, { 8044, 921, 0 }, 1, {8100, 0, 0} },
+   { 2, { 8045, 921, 0 }, 1, {8101, 0, 0} },
+   { 2, { 8046, 921, 0 }, 1, {8102, 0, 0} },
+   { 2, { 8047, 921, 0 }, 1, {8103, 0, 0} },
+   { 2, { 8122, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 913, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 902, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 913, 834, 0 }, 0, {0, 0, 0} },
+   { 3, { 913, 834, 921 }, 0, {0, 0, 0} },
+   { 2, { 913, 921, 0 }, 1, {8115, 0, 0} },
+   { 2, { 8138, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 919, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 905, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 919, 834, 0 }, 0, {0, 0, 0} },
+   { 3, { 919, 834, 921 }, 0, {0, 0, 0} },
+   { 2, { 919, 921, 0 }, 1, {8131, 0, 0} },
+   { 3, { 921, 776, 768 }, 0, {0, 0, 0} },
+   { 3, { 921, 776, 769 }, 0, {0, 0, 0} },
+   { 2, { 921, 834, 0 }, 0, {0, 0, 0} },
+   { 3, { 921, 776, 834 }, 0, {0, 0, 0} },
+   { 3, { 933, 776, 768 }, 0, {0, 0, 0} },
+   { 3, { 933, 776, 769 }, 0, {0, 0, 0} },
+   { 2, { 929, 787, 0 }, 0, {0, 0, 0} },
+   { 2, { 933, 834, 0 }, 0, {0, 0, 0} },
+   { 3, { 933, 776, 834 }, 0, {0, 0, 0} },
+   { 2, { 8186, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 937, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 911, 921, 0 }, 0, {0, 0, 0} },
+   { 2, { 937, 834, 0 }, 0, {0, 0, 0} },
+   { 3, { 937, 834, 921 }, 0, {0, 0, 0} },
+   { 2, { 937, 921, 0 }, 1, {8179, 0, 0} },
+   { 2, { 70, 70, 0 }, 0, {0, 0, 0} },
+   { 2, { 70, 73, 0 }, 0, {0, 0, 0} },
+   { 2, { 70, 76, 0 }, 0, {0, 0, 0} },
+   { 3, { 70, 70, 73 }, 0, {0, 0, 0} },
+   { 3, { 70, 70, 76 }, 0, {0, 0, 0} },
+   { 2, { 83, 84, 0 }, 0, {0, 0, 0} },
+   { 2, { 83, 84, 0 }, 0, {0, 0, 0} },
+   { 2, { 1348, 1350, 0 }, 0, {0, 0, 0} },
+   { 2, { 1348, 1333, 0 }, 0, {0, 0, 0} },
+   { 2, { 1348, 1339, 0 }, 0, {0, 0, 0} },
+   { 2, { 1358, 1350, 0 }, 0, {0, 0, 0} },
+   { 2, { 1348, 1341, 0 }, 0, {0, 0, 0} },
+};
+constexpr uint16_t g_special_case_hash_indices[] = {
```

Comment:
Is this 2nd table necessary? The `g_special_case_mappings` array is 8 bytes of 108 entries or about 1KB. The `g_special_case_hash_indices` array is 2 bytes of 499 entries or about 1KB. So 2KB total. But you have to do two lookups to get to the mapping entry -- first the hash_indices array and then the mappings array. 
If we made a single table of 499 mapping entries this would be about 8 x 499 ~ 4KB. This is still not very much memory but eliminates the 2nd table and all the extra logic around it.
What do you think?

Reply:
Can do.  The special_case_mapping struct is 16 bytes though, so it would be about 8k total.  Still not a big deal.